pkg/vminfo/features.go

```diff
@@ -0,0 +1,192 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package vminfo
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/google/syzkaller/pkg/flatrpc"
+	"github.com/google/syzkaller/pkg/ipc"
+	"github.com/google/syzkaller/pkg/rpctype"
+	"github.com/google/syzkaller/prog"
+)
+
+type Feature struct {
+	Enabled   bool
+	NeedSetup bool
+	Reason    string
+}
+
+type Features map[flatrpc.Feature]Feature
+
+func (features Features) Enabled() flatrpc.Feature {
+	var mask flatrpc.Feature
+	for feat, info := range features {
+		if info.Enabled {
+			mask |= feat
+		}
+	}
+	return mask
+}
+
+func (features Features) NeedSetup() flatrpc.Feature {
+	var mask flatrpc.Feature
+	for feat, info := range features {
+		if info.Enabled && info.NeedSetup {
+			mask |= feat
+		}
+	}
+	return mask
+}
+
+type featureResult struct {
+	id     flatrpc.Feature
+	reason string
+}
+
+func (ctx *checkContext) checkFeatures() {
```

Comment:
Initially I marked features that has executor setup as "don't need to test with an additional program" and they did not produce a test program. But then I removed that for several reasons:
1. Simpler code.
2. Potentially a feature may also have custom code during executor process setup (different from one-time setup), and it's good to test just in case.
3. Almost all (or all?) of these additional programs are deduplicated (e.g. leak/fault/binfmt_misc/swap don't set additional feature flags, so run the same test program with the same flags) and that program matches what we run for sandbox feature check as litmus test (we can disable features, but we still need to ensure that a simple program with no additional features runs, otherwise nothing will work).

Reply:
> Initially I marked features that has executor setup as "don't need to test with an additional program" and they did not produce a test program. 

I meant a slightly different thing. Some features are just not available on a VM (we failed to set them up), and we know that early on. But yes, "don't need to test with an additional program" is an extra signal that could have been used.

If #4762 is merged, then there'd anyway be just a single `Run()` method that would take the argument.

Reply:
Yes, that we can do. But in the current architecture (only possible with #4762 merged).